### PR TITLE
improve download, allow download of big files, works on ff and chrome

### PR DESCRIPTION
**Pull request by kittaakos**
## This is the follow-up of https://github.com/theia-ide/theia/pull/4890

This PR improves the download by letting the native browser handle it
instead of fetching it in the background. This also adds a unique link
which expires in 1 minute. Tested on chrome 73.0 and
firefox 60.5

Signed-off-by: Uni Sayo <unibtc@gmail.com>

### CODE REVIEW

**Comment by kittaakos** [comment]
I have tried it in Gitpod:
 - I could download a file,
 - multiple ones and folders (packed into a `.tar`),
 - download while another download was in progress,
 - also, I could download the entire `Theia` workspace without blocking the UI or experiencing `offline` on the status bar.
 - The `Copy Download Link` worked as well.

@svenefftinge, please review/approve. I am not allowed to do it for a PR with my own changes. Thanks!

File: packages/workspace/src/browser/workspace-commands.ts [hunk 1]
```diff
@@ -152,6 +152,7 @@ export class FileMenuContribution implements MenuContribution {
             commandId: FileDownloadCommands.DOWNLOAD.id,
             order: 'b'
         });
+
```
Comment (akosyakov) [inline packages/workspace/src/browser/workspace-commands.ts]: seems to be only whitespace change

**Comment by akosyakov** [comment]
Generally that's really good, UI is not blocked at all during preparing download.

File: packages/workspace/src/browser/workspace-commands.ts (hunk 1 above)
Comment (kittaakos) [reply packages/workspace/src/browser/workspace-commands.ts]: I see your point, but it was a change in the original code: https://github.com/theia-ide/theia/pull/4890/files#diff-7e70f039e1d4a1a1b544cf48997743e0R149

I did not want to change it, should I?

**Comment by kittaakos** [comment]
> Do we really need `Copy Download Link`? If so i would prefer to have a separate menu item. If i suppose to share it 1m timeout seems to be really small.

[From](https://github.com/theia-ide/theia/pull/4890#issue-269889700) @uniibu:
> And lastly, to be able to use a separate downloader, you will need the download link, so I added a Copy Download Link menu option which works for singe and multiple(tar) files/folders

**Comment by kittaakos** [comment]
> It's strange that i have to press download twice to download.

I did not experience this when I was updating the PR, so I have retested it again; it works. (Tried in Chrome and FF with Gitpod.)

**Comment by akosyakov** [comment]
@kittaakos I meant that i have to click `Download` in the navigator and then in the notification again. For a single file it does not make sense. I'm not sure whether notification makes sense at all, we could simply start downloading and a user can cancel it via the browser.

File: packages/workspace/src/browser/workspace-commands.ts (hunk 1 above)
Comment (akosyakov) [reply packages/workspace/src/browser/workspace-commands.ts]: you can change anything in your commits, but it is up to you, fine to merge like that as well

**Comment by uniibu** [comment]
> > Do we really need `Copy Download Link`? If so i would prefer to have a separate menu item. If i suppose to share it 1m timeout seems to be really small.
> 
> [From](https://github.com/theia-ide/theia/pull/4890#issue-269889700) @uniibu:
> 
> > And lastly, to be able to use a separate downloader, you will need the download link, so I added a Copy Download Link menu option which works for singe and multiple(tar) files/folders

Hi! Its been awhile.. Just want to be clear that, If my memory serves me right, the Download Link acts like a temporary url shortener so you can use it to download big files in parallel using 3rd party downloader and is not meant to be used for sharing. IMO, the 1 minute expiry should be enough for someone to copy it and paste it on their downloader. Even if the link expires as long as the download is already started, it will not be interrupted.

**Comment by kittaakos** [comment]
> use it to download big files in parallel using 3rd party downloader

👍 This. Thank you for the clarification.

**Comment by akosyakov** [comment]
But a special case should not harm the standard scenario? Is it good that a user has to click twice and see notification for a single file?

As for `Copy Download Link` should not it be a separate menu item and then you immediately get a link which you paste somewhere?

**Comment by kittaakos** [comment]
I have made the following changes:
 - From now on there are two separate commands: `Download` and `Copy Download Link`.
 - `Download` remains in the `6_downloadupload` menu group.
 - `Copy Download Link` has been added to the `Edit` > `2_clipboard` main menu group, and to the `5_cutcopypaste` group for the `Navigator` context menu.
 - When `Download` is selected, the download will start automatically right after the prepare phase. Neither further user interaction is required, nor the user will be notified.
 - When `Copy Download Link` is selected, the link will be copied to the clipboard.

Screen Shot 2019-07-10 at 11 28 01
Screen Shot 2019-07-10 at 11 28 08
Screen Shot 2019-07-10 at 11 28 13

**Comment by kittaakos** [comment]
- Copy download link does not always work on FF.
 - If it works, it does not copy the link to the clipboard. I do not know if that was the case before my changes. I have to check.

**Comment by kittaakos** [comment]
> * Copy download link does not always work on FF.

The notification is discarded immediately, it is hard to see for small file(s).
But the link is not copied to the clipboard.

**Comment by kittaakos** [comment]
I do not know how the copy/paste works in FF, but it seems, it is entirely unsupported the way we do: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Interact_with_the_clipboard

**Comment by kittaakos** [comment]
Updating the `copy/paste` functionality is not straightforward, as we do not have type definitions with the current `typescript@3.1.3` version. However, it is already available within the the most recent TypeScript.
See here: https://github.com/microsoft/TypeScript/blob/a4cddd4647154550fe603ec2b65a93cf8b2099dd/lib/lib.dom.d.ts#L10509

**Comment by kittaakos** [comment]
So it seems, in FF one has to actually click on a button to invoke the clipboard-copy, I am going to disable it if the browser is not Chrome.

**Comment by kittaakos** [comment]
PoC: https://jsfiddle.net/kittaakos/ru18g04p/
It works in Chrome but not in FF. One has to remove `display: none` and type into the `input`, lose the focus and see the the text from the `input` was **not** copied to the clipboard.

**Comment by kittaakos** [comment]
Ready for the review.

**Comment by kittaakos** [comment]
Thank you for your help, @uniibu 🥇

**Comment by 502647092** [comment]
@kittaakos @akosyakov 
Copy Download Link Break `Ctrl + C` on edit
temp

**Pull request merged**